text-spacing: text-autospace: handle combining marks

vitorroriz opened the pull request:
#### 41ea61ca1ed39dbad7dd08a24d86d8158a6dabed
<pre>
text-spacing: text-autospace: handle combining marks
<a href="https://bugs.webkit.org/show_bug.cgi?id=284979">https://bugs.webkit.org/show_bug.cgi?id=284979</a>
<a href="https://rdar.apple.com/133797833">rdar://133797833</a>

Reviewed by Ryan Reno.

When a combining mark is present, we should consider its base character
for analyzing the character class for text-autospace.

* LayoutTests/imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-combining-marks-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-combining-marks.html: Added.
* Source/WTF/wtf/text/CharacterProperties.h:
(WTF::isCombiningMark):
* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::InlineItemsBuilder::computeInlineBoxBoundaryTextSpacingsIfNeeded):
(WebCore::Layout::handleTextSpacing):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::baseCharacterFromGraphemeCluster):
(WebCore::Layout::TextUtil::lastBaseCharacterFromText):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.h:
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::adjustGlyphsAndAdvances):

Canonical link: <a href="https://commits.webkit.org/288202@main">https://commits.webkit.org/288202@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/754580e859f2bc85d012cc8362a2fca320c33f9a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82124 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1659 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36091 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/86702 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33162 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84230 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/1705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9487 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/86702 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/21741 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85194 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/1705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74721 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/86702 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/1705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28900 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/31617 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75107 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/1705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29515 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/88150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81179 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9374 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/88150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9562 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70538 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/88150 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17860 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15738 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/14648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/781 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9329 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/14859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103591 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/9176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25136 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/12694 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10976 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-combining-marks.html:
```diff
@@ -0,0 +1,21 @@
+}
+</style>
+<div id="container">
+  <div class="test normal">国&#x301;国&#x301;XX&#x301;国</div>
```
This is good, but you want to make sure you have all ordered combinations of { regular ideographic, combined ideographic, regular ASCII, combined ASCII }; you're missing a few.

fantasai commented on Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
```diff
@@ -717,5 +718,26 @@ bool TextUtil::hasPositionDependentContentWidth(StringView textContent)
+    StringView* lastCluster { nullptr };
+    auto graphemeClusters =  string.graphemeClusters();
+    for (auto cluster : graphemeClusters)
+        lastCluster = &cluster;
```
Analyzing grapheme clusters for the entire string is quite expensive. Since we're just trying to find the last one, I'm pretty sure you can just walk backwards from the end of the string until you get to a non-combining-mark.

vitorroriz replied:
That's a good point. The reason I did that is because `graphemeClusters` returns a class with `begin` and `end` iterators, so I could not skip directly to the last one. But as you mentioned, we are just interested in the last base character so instead of using grapheme cluster's api here we can walk back the string and find the first code point which is not a combining mark I guess.

fantasai commented on Source/WebCore/platform/graphics/ComplexTextController.cpp:
```diff
@@ -776,12 +776,17 @@ void ComplexTextController::adjustGlyphsAndAdvances()
+                    advance.expand(textAutoSpaceSpacing, 0);
+                }
+
+                previousCharacterClass = characterClass;
```
Not sure it's a good idea to move this step... On the one hand, it packages up this logic nicely in one place; on the other it splits up the "let's capture the current state and move forward one step" part. Could go either way, so I'll leave that to you and your official reviewer. :)

vitorroriz replied:
The reason for that is we are making a glyph walk here, not a code point/grapheme cluster walk. The previous character class is solely used for text-autospace computation. We should not update the previous character class when we passed through a combining mark because the adjacency should be computed between the glyph correspondent to the base character and the next glyph.

rreno reviewed (approved):
Rubber-stamping based on fantasai's review.

vitorroriz committed:
text-spacing: text-autospace: handle combining marks
https://bugs.webkit.org/show_bug.cgi?id=284979
rdar://133797833

Reviewed by Ryan Reno.

When a combining mark is present, we should consider its base character
for analyzing the character class for text-autospace.

* LayoutTests/imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-combining-marks-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/text-autospace/text-autospace-combining-marks.html: Added.
* Source/WTF/wtf/text/CharacterProperties.h:
(WTF::isCombiningMark):
* Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp:
(WebCore::Layout::InlineItemsBuilder::computeInlineBoxBoundaryTextSpacingsIfNeeded):
(WebCore::Layout::handleTextSpacing):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::baseCharacterFromGraphemeCluster):
(WebCore::Layout::TextUtil::lastBaseCharacterFromText):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.h:
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::adjustGlyphsAndAdvances):

Canonical link: https://commits.webkit.org/288202@main
